pkg/cloud/cloud.go

```diff
@@ -765,27 +768,130 @@ func (c *cloud) waitForVolume(ctx context.Context, volumeID string) error {
+}
+
+// ResizeDisk resizes an EBS volume in GiB increments,
+// rouding up to the next GiB if reqSize isn't an even GiB.
+// It returns the new size or a negative number if the size couldn't be determined.
+func (c *cloud) ResizeDisk(ctx context.Context, volumeID string, reqSizeBytes int64) (int64, error) {
+	// If current size isgreater or equal, than there's nothing to be done here
+	request := &ec2.DescribeVolumesInput{
+		VolumeIds: []*string{
+			aws.String(volumeID),
+		},
+	}
+	volume, err := c.getVolume(ctx, request)
+	if err != nil {
+		return -1, err
+	}
+
+	// AWS resizes in chunks of GiB (not GB)
+	reqSizeGiB := util.RoundUpGiB(reqSizeBytes)
+	oldSizeGiB := aws.Int64Value(volume.Size)
+
+	if oldSizeGiB >= reqSizeGiB {
+		err := fmt.Errorf("could not expand volume %q: current size (%d GiB) is greater or equal to requested size (%d GiB)", volumeID, oldSizeGiB, reqSizeGiB)
+		return oldSizeGiB, err
+	}
+
+	// Modify volume
+	req := &ec2.ModifyVolumeInput{
+		VolumeId: aws.String(volumeID),
+		Size:     aws.Int64(reqSizeGiB),
+	}
+
+	output, err := c.ec2.ModifyVolumeWithContext(ctx, req)
+	if err != nil {
+		if isAWSErrorIncorrectModification(err) {
+			m, err2 := c.getVolumeModification(ctx, volumeID)
+			if err2 != nil {
+				return -1, fmt.Errorf("could not get modification state of volume %q: %v", volumeID, err)
+			}
+
+			// Check for idempotency
+			state := aws.StringValue(m.ModificationState)
+			if state == ec2.VolumeModificationStateCompleted || state == ec2.VolumeModificationStateOptimizing {
+				return aws.Int64Value(m.TargetSize), nil
+			}
 		}
+		return oldSizeGiB, fmt.Errorf("could not modify AWS volume %q: %v", volumeID, err)
 	}
 
-	return false
+	m := output.VolumeModification
+	if aws.StringValue(m.ModificationState) == ec2.VolumeModificationStateCompleted || aws.StringValue(m.ModificationState) == ec2.VolumeModificationStateOptimizing {
+		return aws.Int64Value(m.TargetSize), nil
+	}
+
+	// The volume should go through the states "modifying", "optimizing", and finally "completed".
+	// If we get at this point, the volume isn't ready yet, so we wait for the correct state before we return.
+	// TODO: according to AWS docs, you can resize the file system as soon as the volume enters the optimizing state.
+	backoff := wait.Backoff{
+		Duration: 1 * time.Second,
+		Factor:   1.8,
+		Steps:    20,
+	}
+
+	var modVolSizeGiB int64
+	waitErr := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		m, err := c.getVolumeModification(ctx, volumeID)
+		if err != nil {
+			return false, err
+		}
+		if aws.StringValue(m.ModificationState) == ec2.VolumeModificationStateCompleted || aws.StringValue(m.ModificationState) == ec2.VolumeModificationStateOptimizing {
+			modVolSizeGiB = aws.Int64Value(m.TargetSize)
+			return true, nil
+		}
+		return true, nil
+	})
+	if waitErr != nil {
+		// Return negative number to avoid describing the volume
+		// to get the actual volume size after the modification failed.
+		return -1, waitErr
```

Comment:
Can we return 0 instead since there is `waitErr` returned anyway?

Reply:
Fixed